### ISSUE TITLE
How to save only parts of the state_dict()

### ISSUE THREAD

**Issue by CaffreyR** [pytorch/pytorch#86205]
### 🐛 Describe the bug

Hi, I want to save only a small part of the model. 
e.g. A layer requires grad but B layer does not. So I only want to save A layer rather than both A and B . Many thanks!

```
def model(nn.Module):
     ...

model=model()

model.save(model.state_dict())
```

### Versions
```
PyTorch version: 1.13.0.dev20220709
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.4 (arm64)
GCC version: Could not collect
Clang version: 13.1.6 (clang-1316.0.21.2.5)
CMake version: version 3.23.1
Libc version: N/A

Python version: 3.9.10 | packaged by conda-forge | (main, Feb  1 2022, 21:25:34)  [Clang 11.1.0 ] (64-bit runtime)
Python platform: macOS-12.4-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.23.2
[pip3] pytorch-ignite==0.4.9
[pip3] pytorch-lightning==1.6.5
[pip3] torch==1.13.0.dev20220709
[pip3] torchaudio==0.14.0.dev20220603
[pip3] torchmetrics==0.9.2
[pip3] torchsummary==1.5.1
[pip3] torchvision==0.14.0.dev20220708
[conda] numpy                     1.23.2                   pypi_0    pypi
[conda] pytorch-ignite            0.4.9                    pypi_0    pypi
[conda] pytorch-lightning         1.6.5                    pypi_0    pypi
[conda] torch                     1.13.0.dev20220709          pypi_0    pypi
[conda] torchaudio                0.14.0.dev20220603          pypi_0    pypi
[conda] torchmetrics              0.9.2                    pypi_0    pypi
[conda] torchsummary              1.5.1                    pypi_0    pypi
[conda] torchvision               0.14.0.dev20220708          pypi_0    pypi
```

cc @albanD @mruberry @jbschlosser @walterddr @kshitij12345 @saketh-are

**Comment by A-D-Cooper** [comment]
`model.stat_dict()` Gives you a dictionary and 
>state_dict objects are Python dictionaries, they can be easily saved, updated, altered, and restored, adding a great deal of modularity to PyTorch models and optimizers. Note that only layers with learnable parameters (convolutional layers, linear layers, etc.) and registered buffers (batchnorm’s running_mean) have entries in the model’s state_dict.

So if a layer doesn't have learnable parameters, it won't be saved. Also you can loop over the dictionary and save the things you want to save and discard the rest. 

Source: https://pytorch.org/tutorials/recipes/recipes/what_is_state_dict.html

**Comment by CaffreyR** [comment]
Hi, @A-D-Cooper , but I try this code. It also prints all the parameters, and the save file is also exists!
'''
import torch
import torch.nn as nn
import torch.optim as optim
class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(3, 6, 5)
        self.pool = nn.MaxPool2d(2, 2)
        self.conv2 = nn.Conv2d(6, 16, 5)
        self.fc1 = nn.Linear(16 * 5 * 5, 120)
        self.fc2 = nn.Linear(120, 84)
        self.fc3 = nn.Linear(84, 10)

    def forward(self, x):
        x = self.pool(F.relu(self.conv1(x)))
        x = self.pool(F.relu(self.conv2(x)))
        x = x.view(-1, 16 * 5 * 5)
        x = F.relu(self.fc1(x))
        x = F.relu(self.fc2(x))
        x = self.fc3(x)
        return x

net = Net()
# print(net)
for name, param in net.named_parameters():
    param.requires_grad = False

print(net.state_dict())
torch.save(net.state_dict(),'model_weights.pth')
'''

**Comment by A-D-Cooper** [comment]
Hi, even though the `requires_grad` is set to False. These parameters are still considered learnable parameters. Many people set the `requires_grad` to False for experimentation but may set them to True for training thus these would still be considered learnable parameters.

**Comment by CaffreyR** [comment]
Hi @A-D-Cooper , could please give me an example of how to loop over the dictionary and change it? Since all my attempts failed. Many thanks!

**Comment by A-D-Cooper** [comment]
Sure.
`
model = ASLClassification()
for layer in model.state_dict():
    print(layer)

conv1.weight
conv1.bias
conv2.weight
conv2.bias
conv3.weight
conv3.bias
conv4.weight
conv4.bias
fc1.weight
fc1.bias
fc2.weight
fc2.bias
fc3.weight
fc3.bias
`
In the above code, I made an instance of class `ASLClassification()' which inherits from `nn.Module`. It's a CNN for ASL classification. It has 4 convolutional layers and 3 fully connected layers (These are the attributes of model). so the first convolutional layers is named 'conv1', second one is 'conv2' and so on. It's same for full connected layer (named 'fc'). when I iterate over them it gives me the name of those layers and I can access the tensors like this. 
`
for layer in model.state_dict():
    print( model.state_dict()[layer] )   # accessing a dictionary

[[-0.0063,  0.0926, -0.0872],
          [ 0.0488, -0.0046,  0.0274],
          [ 0.0401, -0.0050, -0.0926]],

         [[ 0.1346,  0.0994, -0.0831],
          [ 0.0256,  0.0015, -0.0403],
          [ 0.1080, -0.1230,  0.0264]],

         [[-0.0025,  0.1289,  0.1205],
          [-0.0874, -0.0716,  0.0836],
          [ 0.0433,  0.0333,  0.0282]],
...
`

Or if I want to access say the weights of the first fully connected layer, I can do this.
`
print(model.state_dict()['fc1.weight'])   # Note the name of the layer is a string

tensor([[-7.9111e-03,  6.4364e-04,  1.6153e-03,  ...,  4.2058e-03,
         -7.5719e-03, -4.8803e-03],
        [-9.8684e-03,  6.7006e-03,  5.2319e-03,  ...,  3.7247e-03,
          4.3992e-03, -4.2082e-03],
        [-9.7713e-03, -3.6980e-03,  6.4075e-03,  ...,  5.6477e-03,
         -9.0160e-03,  3.8981e-03],
        ...,
`
So now you can make another dictionary and add all the layers you want to keep. Like this

`
new_dict = {}
layers_to_keep = ['conv1.weight', 'fc1.weight', 'fc1.bias']
for layer in layers_to_keep:
    new_dict[layer] = model.state_dict()[layer]
`
so the name of all the layers you want to keep are in `layers_to_keep` list (as a string) and the `new_dict` contains the weights and biases of the layers in the same format as model.state_dict(). you will have to create the 'layers_to_keep` list manually based on which layers you want to keep.

Hope this helps.

**Comment by CaffreyR** [comment]
Hi @A-D-Cooper , sorry for late response, since the Github did not send me notification. Many thanks for your time! I will try it now!:)
